### ISSUE TITLE
Possible RBAC design problem

### ISSUE THREAD

**Issue by boazrymland** [yiisoft/yii#742]
## General

I've written a forum post that received no replies that explains the issue a bit more [forum post](http://www.yiiframework.com/forum/index.php/topic/31987-rbac-system-problem-when-roles-inherit-from-other-role-with-bizrule/page__p__154105#entry154105). I think this is an important issue and that's why I'm raising this bug.
## Description

The initial RBAC setup is simple - have two roles - guest and authenticated. authenticated inherits all from guest so it adds a few more operations on top of the several operations that guest can do. guest and authenticated roles has a nice bizrule that can make the "defaultRoles", meaning no need to attach specifically to every new user on the system. 
This should be quite a powerful setup. It can be summed up in the following diagram:

```
Authenticated (do D, do E)  [bizrule=return !Yii::app()->getUser()->getIsGuest();]
  ^
  |
Guest (do A, do B, do C) [bizrule=return Yii::app()->getUser()->getIsGuest();
```

Now, all good and promising. But:
CDbAuthManager::checkAccessRecursive() runs in a recursive loop bubbling upward with the "parents" of this authorization item, and runs the bizrule if one is found. If the bizrule returned true - access is granted. If false - an immediate reply of "no access" is determined. 
When run on the scheme above, if the authenticated user tries to execute an operation who is allowed to guest (e.g. "do B"), it will be denied access. Why? Because the bizrule for the guest will deny fail, and a negative answer will be returned.
## So in short, what are you saying Boaz?

Now if I understood it all correctly, here's my conclusions in the form of "the essence of the problem" and a "possible design modification to resolve the problem":
### Essence of the problem

Essentially, the problem is that there are only two types of "answers" that are defined in the interface between the bizrule and checkAccessRecursive(). The two answers are "yes - grant access" and "no - deny access". Actually, its not the bizrule that say this, but since only bool is allowed to be returned from the bizrule, those are in fact the only two possible interpretations of the answers. 
**The problem is a missing third option "I don't know"**. And in a more verbose anwser: **"l don't know  - please continue the recursive check. Maybe someone above me (my parents) will either specifically grant the permission, or specifically deny it"**.
### Possible design modification

If, for example, a value of **NULL** was allowed to be returned from the bizrule (that is, one can return it but it will be cast as false... :) ), we could have the bizrule that is assigned to Guest in the above example be "if (Yii::app()->getUser()->getIsGuest()) {return true;} else {return NULL;}", this "null" value could serve as the third type of answer "I don't know". 
For that, checkAccessRecursive() will need to to be updated as well, naturally. The "if" clause on the bizrule execution line will need to be expanded to evaluate the bizrule, check for true, false, or null. In case of null, probably a mere "continue;" will do. In case of true of false, current behavior should be retained.

Whee, that was a long one, wasn't it? I tried to be precise and well-phrased but couldn't make it shorter :)

**Comment by SonyStone** [comment]
it's looks odd that "Performs access check for the specified user" even if $user  =  null.

so i just added:
`        if ($user === null) {
            return false;
        }`
Will it work fine in the future?

![Image of Yaktocat](http://i.imgur.com/8VuDqwQ.png)

**Comment by samdark** [comment]
No. There will be no fixes for 1.1 except security issues and PHP 7 compatibility since it reached EOL.
